engine/rk56.go

```diff
@@ -39,73 +50,105 @@ func (rk *RK56) Step() {
 	M.normalize()
 	torqueFn(k4)
 
 	// stage 5
 	Time = t0 + (4./5.)*Dt_si
-	cuda.Madd5(m, m0, k1, k2, k3, k4, 1, (-8./5.)*h, (144./25.)*h, (-4.)*h, (16./25.)*h)
+	cuda.Madd5(m, m0, rk.k1, k2, k3, k4, 1, (-8./5.)*h, (144./25.)*h, (-4.)*h, (16./25.)*h)
 	M.normalize()
 	torqueFn(k5)
 
 	// stage 6
 	Time = t0 + (1.)*Dt_si
-	cuda.Madd6(m, m0, k1, k2, k3, k4, k5, 1, (361./320.)*h, (-18./5.)*h, (407./128.)*h, (-11./80.)*h, (55./128.)*h)
+	cuda.Madd6(m, m0, rk.k1, k2, k3, k4, k5, 1, (361./320.)*h, (-18./5.)*h, (407./128.)*h, (-11./80.)*h, (55./128.)*h)
 	M.normalize()
 	torqueFn(k6)
 
 	// stage 7
 	Time = t0
-	cuda.Madd5(m, m0, k1, k3, k4, k5, 1, (-11./640.)*h, (11./256.)*h, (-11/160.)*h, (11./256.)*h)
+	cuda.Madd5(m, m0, rk.k1, k3, k4, k5, 1, (-11./640.)*h, (11./256.)*h, (-11/160.)*h, (11./256.)*h)
 	M.normalize()
 	torqueFn(k7)
 
 	// stage 8
 	Time = t0 + (1.)*Dt_si
-	cuda.Madd7(m, m0, k1, k2, k3, k4, k5, k7, 1, (93./640.)*h, (-18./5.)*h, (803./256.)*h, (-11./160.)*h, (99./256.)*h, (1.)*h)
+	cuda.Madd7(m, m0, rk.k1, k2, k3, k4, k5, k7, 1, (93./640.)*h, (-18./5.)*h, (803./256.)*h, (-11./160.)*h, (99./256.)*h, (1.)*h)
 	M.normalize()
 	torqueFn(k8)
 
 	// stage 9: 6th order solution
 	Time = t0 + (1.)*Dt_si
 	//madd6(m, m0, k1, k3, k4, k5, k6, 1, (31./384.)*h, (1125./2816.)*h, (9./32.)*h, (125./768.)*h, (5./66.)*h)
-	cuda.Madd7(m, m0, k1, k3, k4, k5, k7, k8, 1, (7./1408.)*h, (1125./2816.)*h, (9./32.)*h, (125./768.)*h, (5./66.)*h, (5./66.)*h)
+	cuda.Madd7(m, m0, rk.k1, k3, k4, k5, k7, k8, 1, (7./1408.)*h, (1125./2816.)*h, (9./32.)*h, (125./768.)*h, (5./66.)*h, (5./66.)*h)
 	M.normalize()
 	torqueFn(k2) // re-use k2
 
 	// error estimate
 	Err := cuda.Buffer(3, size)
 	defer cuda.Recycle(Err)
-	cuda.Madd4(Err, k1, k6, k7, k8, (-5. / 66.), (-5. / 66.), (5. / 66.), (5. / 66.))
+	cuda.Madd4(Err, rk.k1, k6, k7, k8, (-5. / 66.), (-5. / 66.), (5. / 66.), (5. / 66.))
 
 	// determine error
 	err := cuda.MaxVecNorm(Err) * float64(h)
 
 	// adjust next time step
 	if err < MaxErr || Dt_si <= MinDt || FixDt != 0 { // mindt check to avoid infinite loop
-		// step OK
-		setLastErr(err)
-		setMaxTorque(k2)
-		NSteps++
-		Time = t0 + Dt_si
-		adaptDt(math.Pow(MaxErr/err, 1./6.))
+		//Passed absolute error. Check relative error...
+		errnorm := cuda.Buffer(1, size)
+		defer cuda.Recycle(errnorm)
+		cuda.VecNorm(errnorm, Err)
+		ddtnorm := cuda.Buffer(1, size)
+		defer cuda.Recycle(ddtnorm)
+		cuda.VecNorm(ddtnorm, k2)
+		maxdm := cuda.MaxVecNorm(k2)
+		fail := 0
+		rlerr := float64(0.0)
+		if maxdm < MinSlope { // Only step using relerr if dmdt is big enough. Overcomes equilibrium problem
+			fail = 0
+		} else {
+			cuda.Div(errnorm, errnorm, ddtnorm) //re-use errnorm
+			rlerr = float64(cuda.MaxAbs(errnorm))
+			fail = 1
+		}
+		if fail == 0 || RelErr <= 0.0 || rlerr < RelErr || Dt_si <= MinDt || FixDt != 0 { // mindt check to avoid infinite loop
+			// step OK
+			setLastErr(err)
+			setMaxTorque(k2)
```

Comment:
It seems that the value used in `setMaxTorque()` is inconsistent between different solvers. [NOTE: this was already the case before this PR]

- Here in RK56 (also RK45dp and RK23), `setMaxTorque()` is set with the torques acting on the final magnetization state (i.e. the `torqueFn()` values **after** the final `cuda.Madd<n>(m, m0, ...)` is performed).
- However, in e.g. RK4 (also Heun and Euler I think?), `setMaxTorque()` is set with the torque acting on the magnetization considered in the last stage of the RK solver (i.e. the `torqueFn()` values **before** the final `cuda.Madd<n>(m, m0, ...)` is performed).

Which of these two possibilities do we consider most correct?